doc/developer/design/20240508_subsource_schemas_after_planning.md

```diff
@@ -0,0 +1,112 @@
+# Restructure + genericize subsource planning
+
+-   Associated:
+    -   https://github.com/MaterializeInc/materialize/issues/20208
+    -   https://github.com/MaterializeInc/materialize/pull/26881
+    -   https://github.com/MaterializeInc/materialize/issues/26999
+
+## The Problems
+
+-   To complete #20208, we have to know the schema of the Kafka topic. This
+    schema is most readily knowable after planning the `SOURCE`.
+-   Subsource planning (generating `CreateSubsourceStatement`) currently happens
+    during purification before planning the `SOURCE`.
+
+In addition to these facts being at odds with one another, it's also a poor
+separation of concerns to do so much planning work during purification, when
+much of the work could be done in planning itself.
+
+This is to say that we _never_ should have generated `CreateSubsourceStatement`s
+as a function of purification as none of the work needs to be done
+asynchronously.
+
+## Success Criteria
+
+-   We generate `CreateSubourceStatement`s _after_ planning the
+    `CreateSourceStatement`. This unblocks the ability to theoretically create
+    subsources for Kafka's output.
+-   We demonstrate, using PostgreSQL sources as an example, that it is possible
+    to separate planning and purification concerns.
+
+## Out of Scope
+
+-   Actually creating subsources for Kafka sources.
+-   Schema migrations
+-   Fully separating concerns w/r/t purification, i.e. I am not planning to
+    address [MySQL source's blending of
+    concerns](https://github.com/MaterializeInc/materialize/issues/26999),
+    though I think that the spirit of this change should include that task.
+-   Supporting `ALTER SOURCE...ADD SUBSOURCE` for MySQL, though that should be
+    relatively simple to do.
+
+## Solution Proposal
+
+1.  Modularize source planning. Source planning will become more complex and the
+    current code structure stretches many people's ability to comprehend what's
+    happening.
+2.  Move planning that occurs during purification into planning proper, such as
+    normalizing `TEXT COLUMN` references. Purification should be limited to only
+    do work that _must_ be performed asynchronously. (Note that _should_ is
```

Comment:
Small note: purification is also (correctly, IMO) used for work that should only be performed once, when the user types the SQL into Materialize, vs on every boot, when the SQL is loaded from the catalog. That sort of work is not always asynchronous.

Reply:
Ah, sure––that's a good call. The issue I have is moreso where the code lives vs. when it is called. Will clarify.

Reply:
What if we split up purification into two stages to make this more explicit? async, and then normalization (or whatever other one-off work exists).

The annoyance here is that there is one-off work which we currently do in purification which very closely resembles _other_ work we have to do post-purification which depends on knowing the catalog state when we rejoin the main coord thread (e.g. we have to know all of a source's subsources and cannot assume that the set available to us in purification is the set that exists when we rejoin the main coord thread).

I guess the reason that I'm harping on this which I haven't clearly articulated is that there is a world in which creating a source and altering a source can use the exact same code paths w/r/t managing the subsources that it offers. I'll amend the design doc to split these things out more neatly and clearly.

Reply:
[Here's what I've had in the back of my mind](https://github.com/MaterializeInc/materialize/pull/26881/files#diff-7271ca5f7922adef08c51331d47c305a1e8fb810e2273a8d4b2b80668bf8e3aeR104). In case it moves it's in the section evocatively titled "Mind Palace."

Reply:
> What if we split up purification into two stages to make this more explicit? async, and then normalization (or whatever other one-off work exists).

That makes sense to me!